File: trip-kit/src/main/java/com/skedgo/android/tripkit/routing/SingleRouteService.java
package com.skedgo.android.tripkit.routing;

import android.support.annotation.NonNull;

import com.skedgo.android.common.model.Query;
import com.skedgo.android.common.model.TripGroup;
import com.skedgo.android.tripkit.RouteOptions;
import com.skedgo.android.tripkit.RouteService;

import java.util.List;

import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subjects.Subject;

/**
 * A decorator of {@link RouteService} that performs only one routing request.
 * For example, if we ask it to route from A to B, and while that request
 * is still progress and later we ask it to route from C to B,
 * then the request A-to-B will be cancelled.
 * Cancellation is invoked asynchronously. That means the execution
 * of the request C-to-D doesn't have to wait for
 * the cancellation of the request A-to-B to be done to get started.
 */
public class SingleRouteService implements RouteService {
  /* toSerialized() to be thread-safe. */
  private final Subject<Void, Void> cancellationSignal =
      PublishSubject.<Void>create().toSerialized();
  private final RouteService routeService;

  public SingleRouteService(RouteService routeService) {
    this.routeService = routeService;
  }

  @NonNull @Override public Observable<TripGroup> routeAsync(@NonNull RouteOptions options) {
    cancellationSignal.onNext(null);
    return routeService.routeAsync(options)
        .takeUntil(cancellationSignal.asObservable());
  }

  @NonNull @Override public Observable<List<TripGroup>> routeAsync(@NonNull Query query) {
    cancellationSignal.onNext(null);
    return routeService.routeAsync(query)
        .takeUntil(cancellationSignal.asObservable());
  }
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/a2brouting/SingleRouteService.kt
package com.skedgo.tripkit.a2brouting

import com.skedgo.android.common.model.Query
import com.skedgo.tripkit.TransitModeFilter
import com.skedgo.tripkit.TransportModeFilter
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

import skedgo.tripkit.routing.TripGroup

/**
 * A decorator of [RouteService] that performs only one routing request.
 * For example, if we ask it to route from A to B, and while that request
 * is still progress and later we ask it to route from C to B,
 * then the request A-to-B will be cancelled.
 * Cancellation is invoked asynchronously. That means the execution
 * of the request C-to-D doesn't have to wait for
 * the cancellation of the request A-to-B to be done to get started.
 */
class SingleRouteService(private val routeService: RouteService) : RouteService {
  /* toSerialized() to be thread-safe. */
  private val cancellationSignal = PublishSubject.create<Any>().toSerialized()

  override fun routeAsync(query: Query,
                          transportModeFilter: TransportModeFilter,
                          transitModeFilter: TransitModeFilter): Observable<List<TripGroup>> {
    cancellationSignal.onNext(Any())
    return routeService.routeAsync(query, transportModeFilter, transitModeFilter)
        .takeUntil(cancellationSignal)
  }
}